### PR TITLE
Implement `take`, `truncate`, `skip` and `drop`

### PR DESCRIPTION
**Pull request by abelbraaksma**
This fixes part of #208, the following signatures are added and implemented (todo: tests)

```f#
static member skip: count: int -> source: TaskSeq<'T> -> TaskSeq<'T>
static member drop: count: int -> source: TaskSeq<'T> -> TaskSeq<'T>
static member take: count: int -> source: TaskSeq<'T> -> TaskSeq<'T>
static member truncate: count: int -> source: TaskSeq<'T> -> TaskSeq<'T>
```

Note that `skip` and `take` raise if there are not enough elements (just like its `Seq` counterparts) and that `truncate` and `drop` do not raise. While `drop` does not exist in `Seq`, I felt it make sense as `truncate` also exists. Maybe it can be added to F# Core?

PS, in case anybody wondered, I chose the name `drop` simply because [it exists in Scala](https://bradcollins.com/2015/09/19/scala-saturday-the-stream-drop-method/). I'm not a fan of Scala per se, but was just looking for existing functions that behave like this one, and indeed, the Scala function does not raise either.

See also discussion here: https://stackoverflow.com/questions/1247788/calling-seq-skip-and-seq-take-in-f

### CODE REVIEW

**Comment by abelbraaksma** [comment]
This is ready, now that the obvious errors are fixed ;)

**Comment by abelbraaksma** [comment]
Added `.fantomasignore` in two places (needed for VS 2022 plugin to work). This fixes a (temporary) bug in Fantomas that raises an error because of `WhileBang`, see https://github.com/fsprojects/fantomas/issues/3019

File: src/.fantomasignore [hunk 1]
```diff
@@ -0,0 +1 @@
+TaskSeqInternal.fs
```
Comment (bartelink) [inline src/.fantomasignore:1]: if you move sln up, can this redundancy be removed?

File: src/FSharp.Control.TaskSeq.Test/Nunit.Extensions.fs [hunk 2]
```diff
@@ -138,11 +138,14 @@ module ExtraCustomMatchers =
...
+    ///
     /// Asserts any exception that exactly matches the given exception <see cref="Type" />.
     /// Async exceptions are almost always nested in an <see cref="AggregateException" />, however, in an
     /// async try/catch in F#, the exception is typically unwrapped. But this is not foolproof, and
     /// in cases where we just call <see cref="Task.Wait" />, and <see cref="AggregateException" /> will be raised regardless.
     /// This assertion will go over all nested exceptions and 'self', to find a matching exception.
+    ///
     /// Function to evaluate MUST return a <see cref="System.Threading.Tasks.Task" />, not a generic
     /// <see cref="Task&lt;'T>" />.
     /// Calls <see cref="Assert.ThrowsAnyAsync&lt;Exception>" /> of xUnit to ensure proper evaluation of async.
```
Comment (bartelink) [inline src/FSharp.Control.TaskSeq.Test/Nunit.Extensions.fs:151]: unwrapping/special casing of AggregateException should not be in here....

File: src/FSharp.Control.TaskSeq/TaskSeq.fsi [hunk 3]
```diff
@@ -725,6 +725,65 @@ type TaskSeq =
     /// <exception cref="T:ArgumentNullException">Thrown when the input task sequence is null.</exception>
     static member filterAsync: predicate: ('T -> #Task<bool>) -> source: TaskSeq<'T> -> TaskSeq<'T>
 
+    /// <summary>
+    /// Returns a task sequence that, when iterated, skips <paramref name="count" /> elements of the
+    /// underlying sequence, and then returns the remainder of the elements. Raises an exception if there are not enough
```
Comment (bartelink) [inline src/FSharp.Control.TaskSeq/TaskSeq.fsi:730]:
```suggestion
    /// underlying sequence, and then yields the remainder. Raises an exception if there are not <paramref name="count" /> items. <br/>
```

File: src/FSharp.Control.TaskSeq/TaskSeq.fsi [hunk 4]
```diff
@@ -725,6 +725,65 @@ type TaskSeq =
...
+    /// elements in the sequence. See <see cref="drop" /> for a version that does not raise an exception.
```
Comment (bartelink) [inline src/FSharp.Control.TaskSeq/TaskSeq.fsi:731]:
```suggestion
    /// elements in the sequence. See <see cref="drop" /> for a version that does not raise an exception. <br />
```

File: src/FSharp.Control.TaskSeq/TaskSeq.fsi [hunk 5]
```diff
@@ -725,6 +725,65 @@ type TaskSeq =
...
+
+
+    /// <summary>
+    /// Returns a task sequence that, when iterated, drops at most <paramref name="count" /> elements of the
+    /// underlying sequence, and then returns the remainder of the elements, if any.
+    /// See <see cref="skip" /> for a version that raises an exception if there
+    /// are not enough elements. See also <see cref="truncate" /> for the inverse of this operation.
+    /// </summary>
+    ///
+    /// <param name="count">The number of items to drop.</param>
```
Comment (bartelink) [inline src/FSharp.Control.TaskSeq/TaskSeq.fsi:751]:
```suggestion
    /// <param name="count">The maximum number of items to drop.</param>
```

File: src/FSharp.Control.TaskSeq/TaskSeqInternal.fs [hunk 6]
```diff
@@ -613,6 +631,103 @@ module internal TaskSeqInternal =
...
+        | Skip ->
+            // don't create a new sequence if count = 0
+            if count = 0 then
+                source
+            else
+                taskSeq {
+                    use e = source.GetAsyncEnumerator CancellationToken.None
+
+                    for _ in 1..count do
+                        let! step = e.MoveNextAsync()
```
Comment (bartelink) [inline src/FSharp.Control.TaskSeq/TaskSeqInternal.fs:651]: is more or ok not a better name?

File: src/FSharp.Control.TaskSeq/TaskSeqInternal.fs [hunk 7]
```diff
@@ -613,6 +631,103 @@ module internal TaskSeqInternal =
...
+            else
+                taskSeq {
+                    use e = source.GetAsyncEnumerator CancellationToken.None
+
+                    let! step = e.MoveNextAsync()
+                    let mutable cont = step
+                    let mutable pos = 0
+
+                    // return items until we've exhausted the seq
+                    // report this line, weird error:
```
Comment (bartelink) [inline src/FSharp.Control.TaskSeq/TaskSeqInternal.fs:717]: todo?

**Review by bartelink — APPROVED** [review]

File: src/FSharp.Control.TaskSeq/TaskSeqInternal.fs (hunk 7 above)
Comment (abelbraaksma) [reply src/FSharp.Control.TaskSeq/TaskSeqInternal.fs:717]: Yeah, I haven't reported that line yet. It is the most confusing error I've ever seen 😆. I'll remove that line before merge. Tx for the review!

File: src/.fantomasignore (hunk 1 above)
Comment (abelbraaksma) [reply src/.fantomasignore:1]: Yeah, I have thought about that. There was a reason why I put the `sln` in the `src` folder, but I can't readily recall. If I change it, and I probably should, I would do that in a separate PR to not pollute this one.

The `.fantomasignore` is going to be temporary anyway.

File: src/FSharp.Control.TaskSeq.Test/Nunit.Extensions.fs (hunk 2 above)
Comment (abelbraaksma) [reply src/FSharp.Control.TaskSeq.Test/Nunit.Extensions.fs:151]: Why not? It is a helper function for testing. The `AggregateException` is a result of how exceptions are aggregated from threads and doing that by hand in each test and each scenario is just a lot of work.

All we are after is if the (nested or not) code raises the said exception. This is used in `should throw typeof<..>` scenarios.

File: src/FSharp.Control.TaskSeq/TaskSeq.fsi (hunk 3 above)
Comment (abelbraaksma) [reply src/FSharp.Control.TaskSeq/TaskSeq.fsi:730]: Tx, I like it. Updated.

File: src/.fantomasignore (hunk 1 above)
Comment (bartelink) [reply src/.fantomasignore:1]: Generally the reason I run into is that if you have a `.proj` and a `.sln` in the same folder, dotnet build and things start complaining about ambiguity
Agree separate PR makes sense for that

File: src/FSharp.Control.TaskSeq/TaskSeq.fsi [hunk 8]
```diff
@@ -725,6 +725,65 @@ type TaskSeq =
...
+    /// elements in the sequence. See <see cref="drop" /> for a version that does not raise an exception.
+    /// See also <see cref="take" /> for the inverse of this operation.
+    /// </summary>
+    ///
+    /// <param name="count">The number of items to skip.</param>
+    /// <param name="source">The input task sequence.</param>
+    /// <returns>The resulting task sequence.</returns>
+    /// <exception cref="T:ArgumentNullException">Thrown when the input task sequence is null.</exception>
+    /// <exception cref="T:ArgumentException">Thrown when <paramref name="count" /> is less than zero.</exception>
+    /// <exception cref="T:InvalidOperationException">Thrown when count exceeds the number of elements in the sequence.</exception>
```
[the remark this answers is not in the document]
Comment (abelbraaksma) [reply src/FSharp.Control.TaskSeq/TaskSeq.fsi:740]: I ended up combining this and prev line, as this throws `ArgumentException` and not `InvalidOperationException`.

File: src/FSharp.Control.TaskSeq/TaskSeq.fsi (hunk 4 above)
Comment (abelbraaksma) [reply src/FSharp.Control.TaskSeq/TaskSeq.fsi:731]: The `<br />` is not rendered (bug in F#). An empty line is, but adds a thin paragraph spacer.

File: src/FSharp.Control.TaskSeq/TaskSeq.fsi (hunk 5 above)
Comment (abelbraaksma) [reply src/FSharp.Control.TaskSeq/TaskSeq.fsi:751]: Tx, nice catch. Done.

File: src/FSharp.Control.TaskSeq/TaskSeqInternal.fs (hunk 6 above)
Comment (abelbraaksma) [reply src/FSharp.Control.TaskSeq/TaskSeqInternal.fs:651]: Updated. I should probably go over the code in general and find some kind of standard naming for often-used variable types.

File: src/.fantomasignore (hunk 1 above)
Comment (abelbraaksma) [reply src/.fantomasignore:1]: yeah, but that's not the case here. There is no `proj` and `sln` in the same folder. It is just one down from the root, but structure is standard.

Having them an extra directory apart gives issues in Visual Studio when you add a new project, which is will add by default to the root dir (in a project dir), but not under `src`. Which is probably why i did it this way.

Neither approach would solve the issue with Fantomas, btw.

**Comment by abelbraaksma** [comment]
Thanks for the review, @bartelink. I'll merge this in as soon as it passes the tests.

File: src/.fantomasignore (hunk 1 above)
Comment (bartelink) [reply src/.fantomasignore:1]: >Neither approach would solve the issue with Fantomas, btw.

Then don't do that :D

But yes, having a separate `src` vs `tests` split is obv only actually useful once you stop having 1:1 sets of X and X.Tests and/or something that you'd put in a Directory.props file which would vary between tests and src (and/or enable you not to put redundant stuff at top level)

>Neither approach would solve the issue with Fantomas, btw.

pity, and I guess thats the main concern

**Comment by abelbraaksma** [comment]
I sorted out the issue with the test reports not showing. Reported here: https://github.com/dorny/test-reporter/issues/343 (see my comments). See also this thread: https://github.com/actions/upload-artifact/issues/472

File: src/.fantomasignore (hunk 1 above)
Comment (abelbraaksma) [reply src/.fantomasignore]: Well, good news, in part, as it was suggested to use Fantomas alpha release (see https://github.com/fsprojects/fantomas/issues/3019#issuecomment-1863079619) to at least get rid of the `.fantomasignore` file again.

We still have the double manifest, but at least we can now format `WhileBang` properly.

Comment (abelbraaksma) [reply src/.fantomasignore]: I reverted that change and rebased.

**Pull request merged**